torch/distributed/pipelining/microbatch.py

```diff
@@ -115,6 +116,111 @@ class _Replicate:
     pass
 
 
+def _split_block_mask(
+    block_mask: BlockMask,
+    spec: TensorChunkSpec,
+    num_chunks: int,
+) -> list[BlockMask]:
+    """Given a block mask, and a chunking spec, split the block mask.
+
+    Args:
+        block_mask: Block mask to split
+        spec: Chunking spec
+        num_chunks: Number of chunks to split the block mask into
+
+    Returns:
+        chunk_block_masks: List of chunked block masks
+    """
+
+    assert block_mask.kv_num_blocks.size(0) >= num_chunks, (
+        "Block mask has fewer batch size than the number of chunks. "
+    )
+
+    kv_num_blocks_chunks = torch.tensor_split(
+        block_mask.kv_num_blocks, num_chunks, spec.split_dim
+    )
+    kv_indices_chunks = torch.tensor_split(
+        block_mask.kv_indices, num_chunks, spec.split_dim
+    )
+    full_kv_num_blocks_chunks = (
+        torch.tensor_split(block_mask.full_kv_num_blocks, num_chunks, spec.split_dim)
+        if block_mask.full_kv_num_blocks is not None
+        else [None] * num_chunks
+    )
+    full_kv_indices_chunks = (
+        torch.tensor_split(block_mask.full_kv_indices, num_chunks, spec.split_dim)
+        if block_mask.full_kv_indices is not None
+        else [None] * num_chunks
+    )
+
+    chunk_block_masks = []
+    batch_offset = 0
+    for chunk_idx in range(num_chunks):
+
+        def create_mask_mod(idx):
+            def mask_mod_wrapper(b, h, q_idx, kv_idx):
+                b_offset = torch.full_like(b, idx)
+                return block_mask.mask_mod(b + b_offset, h, q_idx, kv_idx)
+
+            return mask_mod_wrapper
+
+        # `from_kv_blocks` is strictly faster than `create_block_mask`.
```

Comment:
nit, I dont think you need to say this,  if you have the final tensors you dont need to go through and apply vmap + mask-mod to build them from scratch. So its more like create_block_mask is a factory and it ends up calling from_kv_blocks which is the the base constructor